### PR TITLE
Input and Output Schema

### PR DESCRIPTION
**Pull request by schrockn**
This is a continuing refactor of the type system. Whenever one wants to implement a type with input and output config schemas, you implement subclasses of InputSchema and OutputSchema, respectively.

### CODE REVIEW

File: python_modules/dagster/dagster/core/types/materializable.py [hunk 1]
```diff
@@ -1,17 +1,11 @@
...
-class Materializeable(object):
-    def define_materialization_config_schema(self):
-        check.failed('must implement')
-
-    def materialize_runtime_value(self, _config_spec, _runtime_value):
-        check.failed('must implement')
+import pickle
 
 
 class FileMarshalable:
```
Comment (mgasner) [inline python_modules/dagster/dagster/core/types/materializable.py]: `Pickleable`? also, this plays a little fast and loose with path / file language as written

File: python_modules/dagster/dagster/core/types/field.py [hunk 2]
```diff
@@ -207,3 +207,24 @@ def __init__(self):
             )
 
     return _Dict
+
+
+def Selector(fields):
```
Comment (mgasner) [inline python_modules/dagster/dagster/core/types/field.py]: beautiful

File: python_modules/dagster/dagster/core/types/config_schema.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
+from dagster import check
+
+from .config import resolve_to_config_type
+
+
+class InputSchema:
```
Comment (mgasner) [inline python_modules/dagster/dagster/core/types/config_schema.py]: now may not be the time but i think we should be using abc for this https://docs.python.org/3/library/abc.html

File: python_modules/dagster/dagster/core/types/builtin_config_schemas.py [hunk 4]
```diff
@@ -13,13 +14,14 @@ def define_path_dict_field():
...
 
-    class _MaterializeableBuiltinScalarConfigSchema(ConfigSelector):
-        def __init__(self):
-            super(_MaterializeableBuiltinScalarConfigSchema, self).__init__(
-                name=scalar_name + '.MaterializationSchema',
-                description='Materialization schema for scalar ' + scalar_name,
-                fields={'json': define_path_dict_field()},
-            )
+    schema_cls = NamedSelector(
+        scalar_name + '.MaterializationSchema', {'json': define_path_dict_field()}
```
Comment (mgasner) [inline python_modules/dagster/dagster/core/types/builtin_config_schemas.py]: I think the time is coming when we will have to revisit the system-generated names

Comment (schrockn) [reply python_modules/dagster/dagster/core/types/builtin_config_schemas.py]: yup

**Pull request merged**